rudder-web/src/main/scala/com/normation/rudder/web/services/SrvGrid.scala

```diff
@@ -132,8 +136,95 @@ class SrvGrid(
 
     JsRaw(s"""createNodeTable("${tableId}",${data.json.toJsCmd},"${S.contextPath}",${refresh});""")
 
-   }
+  }
+
+ // Ajax call back to get compliance details
+  private[this] def ajaxCompliance(tableId : String, future : Future[Box[Map[NodeId,Option[ComplianceLevel]]]]) : JsCmd = {
+    SHtml.ajaxInvoke( () => {
+      // Is my future completed ?
+      if( future.isCompleted ) {
+        // Yes wait for result
+        Await.result(future,scala.concurrent.duration.Duration.Inf) match {
+          case Full(compliances) =>
+            val bars  = {
+              for { (nodeId, optCompliance) <- compliances } yield {
+                optCompliance match {
+                  case None =>
+                    s"""
+                    $$("#compliance-bar-${nodeId.value}").parent().css("text-decoration", "none");
+                    $$("#compliance-bar-${nodeId.value}").html('<div class="tw-bs"><span class="text-center text-muted">no data available</span></div>');
+                    """
+                  case Some(compliance) =>
+                    val array = JsArray (
+                        JE.Num(compliance.pc_notApplicable)
+                      , JE.Num(compliance.pc_success)
+                      , JE.Num(compliance.pc_repaired)
+                      , JE.Num(compliance.pc_error)
+                      , JE.Num(compliance.pc_pending)
+                      , JE.Num(compliance.pc_noAnswer)
+                      , JE.Num(compliance.pc_missing)
+                      , JE.Num(compliance.pc_unexpected)
+                    )
+                    s"""
+                    $$("#compliance-bar-${nodeId.value}").html(buildComplianceBar(${array.toJsCmd}));
+                    nodeCompliances['${nodeId.value}'] = ${array.toJsCmd};
+                    """
+                }
+              }
+            }
+            JsRaw(
+              s"""
+              ${bars.mkString(";")}
+              resortTable("${tableId}")
+              """
+            )
+
+          case eb : EmptyBox =>
+            val error = eb ?~! "error while fetching compliances"
+            logger.error(error.messageChain)
+            Alert(error.messageChain)
+        }
+      } else {
+        After(500,ajaxCompliance(tableId,future))
+      }
+    } )
+
+  }
 
```

Comment:
That code and the other belove are almost line for line what is in RuleGrid. Could you factor out that code and use it in the two grid - else maintenance will be horrible.